java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java

```diff
@@ -0,0 +1,582 @@
+import org.apache.arrow.flight.FlightClient;
+import org.apache.arrow.flight.FlightClientMiddleware;
+import org.apache.arrow.flight.FlightEndpoint;
+import org.apache.arrow.flight.FlightInfo;
+import org.apache.arrow.flight.FlightRuntimeException;
+import org.apache.arrow.flight.FlightStream;
+import org.apache.arrow.flight.Location;
+import org.apache.arrow.flight.auth2.BearerCredentialWriter;
+import org.apache.arrow.flight.auth2.ClientBearerHeaderHandler;
+import org.apache.arrow.flight.auth2.ClientIncomingAuthHeaderMiddleware;
+import org.apache.arrow.flight.client.ClientCookieMiddleware;
+import org.apache.arrow.flight.grpc.CredentialCallOption;
+import org.apache.arrow.flight.sql.FlightSqlClient;
+import org.apache.arrow.flight.sql.impl.FlightSql.SqlInfo;
+import org.apache.arrow.flight.sql.util.TableRef;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.util.AutoCloseables;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.apache.calcite.avatica.Meta.StatementType;
+
+/**
+ * A {@link FlightSqlClient} handler.
+ */
+public final class ArrowFlightSqlClientHandler implements AutoCloseable {
+
+  private final FlightSqlClient sqlClient;
+  private final Set<CallOption> options = new HashSet<>();
+
+  ArrowFlightSqlClientHandler(final FlightSqlClient sqlClient,
+                              final Collection<CallOption> options) {
+    this.options.addAll(options);
+    this.sqlClient = Preconditions.checkNotNull(sqlClient);
+  }
+
+  /**
+   * Creates a new {@link ArrowFlightSqlClientHandler} from the provided {@code client} and {@code options}.
+   *
+   * @param client  the {@link FlightClient} to manage under a {@link FlightSqlClient} wrapper.
+   * @param options the {@link CallOption}s to persist in between subsequent client calls.
+   * @return a new {@link ArrowFlightSqlClientHandler}.
+   */
+  public static ArrowFlightSqlClientHandler createNewHandler(final FlightClient client,
+                                                             final Collection<CallOption> options) {
+    return new ArrowFlightSqlClientHandler(new FlightSqlClient(client), options);
+  }
+
+  /**
+   * Gets the {@link #options} for the subsequent calls from this handler.
+   *
+   * @return the {@link CallOption}s.
+   */
+  private CallOption[] getOptions() {
+    return options.toArray(new CallOption[0]);
+  }
+
+  /**
+   * Makes an RPC "getStream" request based on the provided {@link FlightInfo}
+   * object. Retrieves the result of the query previously prepared with "getInfo."
+   *
+   * @param flightInfo The {@link FlightInfo} instance from which to fetch results.
+   * @return a {@code FlightStream} of results.
+   */
+  public List<FlightStream> getStreams(final FlightInfo flightInfo) {
+    return flightInfo.getEndpoints().stream()
+        .map(FlightEndpoint::getTicket)
+        .map(ticket -> sqlClient.getStream(ticket, getOptions()))
```

Comment:
I think this behavior is incorrect, and it is causing problems for DataFusion's implementation. It looks like this code re-uses the existing `sqlClient`, ignoring the value of `flightInfo.enpoints[x].locations[y].uri` (which may be different than the original URI the `sqlClient` connected to, in the case of DataFusion, the `sqlClient` would be connected to the scheduler at one URI, and the data would be held on one or more executor nodes at different URIs).

I can attempt a PR if you agree with my hypothesis. I'm new to all this, so I could be way off base. And thank you for putting in the massive amount of work necessary to make this BTW!

Reply:
You are right, we actually merged a clarification for this behavior a little while ago. https://github.com/apache/arrow/pull/12636

Note we're still finalizing the IP clearance of this PR (but I assume it will go through one way or another)

Reply:
Glad to hear it's been fixed. Is there a current, long-running rep & branch we can track until this gets merged into arrow proper?

Reply:
This is the repo/branch being used for the IP clearance so I would say you're already looking at it!

My plan is once the clearance goes through, this will get merged into a branch on apache/arrow (the PR merge target here, flight-sql-jdbc), then we'll merge master into the branch, then finally do a PR into master (and review the code and such).

Reply:
Oh, I think I understand now:

> merged a clarification for this behavior

The behavior has been clarified, but the code to which I linked has not been corrected according to the now clarified expected behavior?

If that's the case, I'll work on a fix and PR this branch.

Reply:
Yes, you got it. We clarified the spec, but haven't necessarily updated everything else

Reply:
This works, but I'm interested in feedback about how to do it the correct way: https://github.com/rafael-telles/arrow/pull/42

Reply:
I see James followed up already. Thanks for taking a look!